internal/tasks/account_deletion.go

```diff
@@ -0,0 +1,213 @@
+			return err
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	if len(nextManifests) > 0 {
+		// deletion was not completed yet -> check if we need to do something on our side
+		logg.Info("cleaning up account %q: need to delete %d manifests in this cycle", account.Name, len(nextManifests))
+		for _, rm := range nextManifests {
+			parsedDigest, err := digest.Parse(rm.Digest)
+			if err != nil {
+				return fmt.Errorf("while deleting manifest %q in repository %q: could not parse digest: %w",
+					rm.Digest, rm.RepositoryName, err)
+			}
+			repo, err := keppel.FindRepository(j.db, rm.RepositoryName, account.Name)
+			if err != nil {
+				return fmt.Errorf("while deleting manifest %q in repository %q: could not find repository in DB: %w",
+					rm.Digest, rm.RepositoryName, err)
+			}
+			err = j.processor().DeleteManifest(ctx, account.Reduced(), *repo, parsedDigest, actx)
+			if err != nil {
+				return fmt.Errorf("while deleting manifest %q in repository %q: %w",
+					rm.Digest, rm.RepositoryName, err)
+			}
+		}
+	}
+
+	// delete all repos (and therefore, all blob mounts), so that blob sweeping can immediately take place
+	_, err = j.db.Exec(deleteAccountReposQuery, account.Name)
+	if err != nil {
+		return err
+	}
+
+	// can only delete account when all blobs have been deleted
+	blobCount, err := j.db.SelectInt(deleteAccountCountBlobsQuery, account.Name)
+	if err != nil {
+		return err
+	}
+	if blobCount > 0 {
+		// make sure that blob sweep runs immediately
+		// TODO: how to prevent resetting time stamp if already set?
+		_, err := j.db.Exec(deleteAccountMarkAllBlobsForDeletionQuery, account.Name, j.timeNow())
+		if err != nil {
+			return err
+		}
+		_, err = j.db.Exec(deleteAccountScheduleBlobSweepQuery, account.Name, j.timeNow())
+		if err != nil {
+			return err
+		}
+		logg.Info("cleaning up managed account %q: waiting for %d blobs to be deleted", account.Name, blobCount)
+		return nil
+	}
+
+	// start deleting the account in a transaction
+	tx, err := j.db.Begin()
+	if err != nil {
+		return err
+	}
+	defer sqlext.RollbackUnlessCommitted(tx)
+	_, err = tx.Delete(&account)
+	if err != nil {
+		return err
+	}
+
+	// before committing the transaction, confirm account deletion with the
+	// storage driver and the federation driver
+	err = j.sd.CleanupAccount(ctx, account.Reduced())
+	if err != nil {
+		return fmt.Errorf("while cleaning up storage for account: %w", err)
+	}
+	err = j.fd.ForfeitAccountName(ctx, account)
+	if err != nil {
+		return fmt.Errorf("while cleaning up name claim for account: %w", err)
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return err
+	}
+
+	if userInfo := actx.UserIdentity.UserInfo(); userInfo != nil {
+		j.auditor.Record(audittools.EventParameters{
+			Time:       j.timeNow(),
+			Request:    actx.Request,
+			User:       userInfo,
+			ReasonCode: http.StatusOK,
+			Action:     cadf.DeleteAction,
+			Target:     processor.AuditAccount{Account: account},
+		})
+	}
```

Comment:
The user's DELETE already created an audit event. Since we're acting on behalf of the user here, a separate audit event is not necessary.